src/operator/communication/shard.rs

```diff
@@ -0,0 +1,284 @@
+        // and hence sharding is a no-op.  In this case, we simply return the
+        // input stream.  This allows us to use `shard` unconditionally without
+        // incurring any overhead in the single-threaded case.
+        self.shard_generic().unwrap_or_else(|| self.clone())
+    }
+
+    /// Like [`Self::shard`], but can assemble the results into any output batch
+    /// type `OB`.
+    ///
+    /// Returns `None` when the circuit is not running inside a multithreaded
+    /// rutime or is running in a runtime with a single worker thread.
+    pub fn shard_generic<OB>(&self) -> Option<Stream<Circuit<P>, OB>>
+    where
+        OB: Batch<Key = IB::Key, Val = IB::Val, Time = (), R = IB::R> + Clone + Send + 'static,
+    {
+        Runtime::runtime().and_then(|runtime| {
+            let num_workers = runtime.num_workers();
+
+            if num_workers == 1 {
+                None
+            } else {
+                let output = self
+                    .circuit()
+                    .cache_get_or_insert_with(
+                        ShardId::new((
+                            self.origin_node_id().clone(),
+                            sharding_policy(self.circuit()),
+                        )),
+                        move || {
+                            // As a minor optimization, we reuse this array across all invocations
+                            // of the sharding operator.
+                            let mut builders = Vec::with_capacity(runtime.num_workers());
+                            let (sender, receiver) = self.circuit().new_exchange_operators(
+                                &runtime,
+                                Runtime::worker_index(),
+                                move |batch: IB, batches: &mut Vec<OB>| {
+                                    Self::shard_batch(&batch, num_workers, &mut builders, batches);
+                                },
+                                |trace: &mut Spine<OB>, batch: OB| trace.insert(batch),
+                            );
+                            // Is `consolidate` always necessary? Some (all?) consumers may be happy
+                            // working with traces.
+                            self.circuit()
+                                .add_exchange(sender, receiver, self)
+                                .consolidate()
+                        },
+                    )
+                    .clone();
+                Some(output)
+            }
+        })
+    }
+
+    /// Collect all shards of a stream at the same worker.
+    ///
+    /// The output stream in `receiver_worker` will contain a union of all
+    /// input batches across all workers.  The output streams in all other
+    /// workers will contain empty batches.
+    pub fn gather(&self, receiver_worker: usize) -> Stream<Circuit<P>, IB>
+    where
+        IB: Batch + Send,
+    {
+        match Runtime::runtime() {
```

Comment:
so circuits are deterministic in terms of the zset values produced, but not necessarily in the representation of the zsets?
In particular, gather's output may be dependent on scheduling order?
was this true before?

Reply:
Because we call `consolidate`, the output ends up converted to the canonical representation.  But even if it didn't the trace would still yield key/value pairs in the canonical order. So it should all be deterministic.